Preferences fixes

Be-ing opened the pull request:
* Split bottom sections of Interface preference pane into new Deck pane
  * Fix Interface and Deck panes so settings are only applied when pressing Apply button
  * Fix loading of rate range from empty profile https://bugs.launchpad.net/mixxx/+bug/1676140

Be-ing committed:
DlgPrefControls: remove "(Experimental)" label from skin scaling

Be-ing committed:
DlgPrefControls: don't apply changes until Apply button is pressed

Only did this for the top section so far. The rest will be split into a
new Decks pane.

Be-ing committed:
split Interface preference pane into Interface and Decks

Be-ing committed:
DlgPrefDeck: detangle rate range combobox from CO value

so that Apply button must be pressed for changed setting to take effect

TODO: do this for the rest of the preferences in DlgPrefDeck

Be-ing committed:
DlgPrefDeck: detangle rate inversion checkbox from CO

until Apply button is pressed

Be-ing committed:
code cleanup (use enum classes instead of magic ints)

Be-ing committed:
change "rateRange" CO to a ControlPotmeter to set limits on it

to prevent recurrence of bugs like
https://bugs.launchpad.net/mixxx/+bug/1631393 when rateRange was 0

Be-ing committed:
DlgPrefDeck: detangle playing track protection checkbox from config value

until Apply button is pressed

Be-ing committed:
DlgPrefDeck: detangle load tracks to cue point checkbox from ConfigValue

until Apply button is pressed

Be-ing committed:
DlgPrefDeck: detangle cue mode selection buttons from ConfigValue

until Apply button is pressed

Be-ing committed:
DlgPrefDeck: detangle rate adjustment widgets from ConfigValues

until Apply button is pressed

Be-ing committed:
DlgPrefDeck: detangle track time setting from CO

until Apply button is pressed

Be-ing commented:
Yikes there were a lot of dusty cobwebs in that code! I didn't take care of all of them, but it should be more manageable now. Please test and make sure nothing broke.

Copying @ywwg's [checklist](https://github.com/mixxxdj/mixxx/pull/527#issuecomment-158114075) for testing preference changes:
  * [ ] Restore defaults should work
  * [ ] change some values, hit cancel, reenter pane. Changes should have been cancelled
  * [ ] change some values, hit apply, close and reenter pane. Changes should have been accepted
  * [ ] change some values, apply, see that Mixxx actually obeys the changes.
  * [ ] change some values, apply, exit mixxx, restart. Changes should still be reflected.

Be-ing commented on src/preferences/dialog/dlgpreferences.cpp:
```diff
@@ -170,6 +173,12 @@ void DlgPreferences::createIcons() {
     m_pControlsButton->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled);
 
+    m_pDecksButton = new QTreeWidgetItem(contentsTreeWidget, QTreeWidgetItem::Type);
+    m_pDecksButton->setIcon(0, QIcon(":/images/preferences/ic_preferences_decks.png"));
```
Note that this icon does not exist yet. @ronso0 or @esbrandt would you like to make it?

ronso0 replied:
Something like a Play icon should do.

![ic_preferences_decks](https://user-images.githubusercontent.com/5934199/32701107-e11f6320-c7cf-11e7-9e3f-071886ced0c8.png)

![ic_preferences_decks2](https://user-images.githubusercontent.com/5934199/32701055-60cc3d38-c7cf-11e7-8b5a-4edd6d1ad636.png)

ronso0 replied:
What's your preference? I like the boxed Play more because I asscociate it with a container rather than behaviour

ronso0 replied:
You can take them from the comment, they are not altered/compressed.

ronso0 commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -0,0 +1,623 @@
+/***************************************************************************
+                           dlgprefcontrols.cpp  -  description
```
dlgprefdeck.cpp ?

Be-ing replied:
Good catch, thanks. That comment could be removed entirely.

daschuer commented:
If you change to auto scaling the default skin is loaded and not the configured skin. 

How about moving the waveform options into root?

Be-ing commented:
> If you change to auto scaling the default skin is loaded and not the configured skin.

I'll look into this.

> How about moving the waveform options into root?

Yep, I am planning on it.

sblaisot commented:
Can you please provide before/after screenshots in UI related PRs? this way, one doesn't have to compile/install a version including this PR to see what changed.

ronso0 commented:
[Bug 1731774](https://bugs.launchpad.net/mixxx/+bug/1731774) could be fixed here as well.
Do you plan to do so? If yes, do we have vector sources for those images somewhere?
I can help with that, it's a matter of minutes once we agreed on an appropriate bg color.

Pegasus-RPG commented:
I second the inclusion of before & after screen shots.

Be-ing commented:
Not much has changed with the GUI. I'm mostly looking for people to test to make sure nothing broke.

Before:
![image](https://user-images.githubusercontent.com/9455094/32872914-e45103d8-ca4e-11e7-8bda-f7e154f4da08.png)


After:
![image](https://user-images.githubusercontent.com/9455094/32872723-c04d2864-ca4d-11e7-9c82-f2bca522acf2.png)
![image](https://user-images.githubusercontent.com/9455094/32872747-e0c0f350-ca4d-11e7-86e6-100edaf91179.png)

ronso0 commented:
Looks nice!

Now that the Interface section is cleaned up, what about displaying the skin `<description>`
below skin drop-down like we have for controllers?

Be-ing committed:
fix reverting to default skin when opening Preferences without pressing Apply

somehow DlgPrefInterface::slotApply is shown when closing the
Preferences window even if the Interface pane is never shown and the
Okay button is used to close the window.

Also, configuration values should never be set outside of the
preferences code!

Be-ing commented on src/skin/skinloader.cpp:
```diff
@@ -71,8 +71,6 @@ QString SkinLoader::getConfiguredSkinPath() const {
             configSkin = getDefaultSkinName();
         }
-        m_pConfig->set(ConfigKey("[Config]", "ResizableSkin"),
-                       ConfigValue(configSkin));
```
It should not be possible to set configuration values outside of the preferences code! The whole preferences system could really use a thorough cleanup. It's really brittle.

Be-ing committed:
move Waveform preferences out to equal tree level as other prefs

Be-ing committed:
cleanup variable names in DlgPreferences

Be-ing committed:
reorder preferences pages in list

This order is somewhat arbitrary. I am trying to put it in order
of what I think is most important for users to configure first.

Be-ing committed:
add icon for Decks Preferences

Thanks @ronso0 !

Be-ing commented:
Did some bug hunting and code cleanup. I also rearranged the order of the preference pages, roughly in order of what is most important for users to set up after installing Mixxx (which is somewhat arbitrary):
![image](https://user-images.githubusercontent.com/9455094/32998581-fb1493f8-cd61-11e7-8dcc-7f9adb88e580.png)

sblaisot commented:
Appveyor tels we now have 10 failed test under windows

```
[00:32:49] [  PASSED  ] 456 tests.
[00:32:49] [  FAILED  ] 10 tests, listed below:
[00:32:49] [  FAILED  ] EngineBufferE2ETest.BasicProcessingTest
[00:32:49] [  FAILED  ] EngineBufferE2ETest.KeylockReverseTest
[00:32:49] [  FAILED  ] EngineSyncTest.InternalMasterSetFollowerSliderMoves
[00:32:49] [  FAILED  ] EngineSyncTest.RateChangeTest
[00:32:49] [  FAILED  ] EngineSyncTest.RateChangeTestWeirdOrder
[00:32:49] [  FAILED  ] EngineSyncTest.RateChangeTestOrder3
[00:32:49] [  FAILED  ] EngineSyncTest.FollowerRateChange
[00:32:49] [  FAILED  ] EngineSyncTest.InternalRateChangeTest
[00:32:49] [  FAILED  ] EngineSyncTest.MasterStopSliderCheck
[00:32:49] [  FAILED  ] EngineSyncTest.SyncToNonSyncDeck
[00:32:49] 
[00:32:49] 10 FAILED TESTS
```

daschuer commented on src/mixxx.cpp:
```diff
@@ -366,13 +366,13 @@ void MixxxMainWindow::initialize(QApplication* pApp, const CmdlineArgs& args) {
-                                                           m_pEffectsManager,
-                                                           m_pRecordingManager))) {
+    if (!(m_pWidgetParent = m_pSkinLoader->loadConfiguredSkin(this, m_pKeyboard,
+                                                              m_pPlayerManager,
```
Thats why the parenthese alignment ... to much noise ;-)

Be-ing replied:
I see what you mean here. I don't think we have a specific rule in the style guide about how to indent function calls with many parameters. How should it be written?

daschuer replied:
You can keep it like it is. It was only a kidding comment. 

This is the way Eclipse auto fomats it:
```C++
QWidget* SkinLoader::loadConfiguredSkin(QWidget* pParent,
        KeyboardEventFilter* pKeyboard, PlayerManager* pPlayerManager,
        ControllerManager* pControllerManager, Library* pLibrary,
        VinylControlManager* pVCMan, EffectsManager* pEffectsManager,
        RecordingManager* pRecordingManager) {
```

This is my favourite:
```C++
QWidget* SkinLoader::loadConfiguredSkin(
        QWidget* pParent,
        KeyboardEventFilter* pKeyboard, 
        PlayerManager* pPlayerManager,
        ControllerManager* pControllerManager, 
        Library* pLibrary,
        VinylControlManager* pVCMan, 
        EffectsManager* pEffectsManager,
        RecordingManager* pRecordingManager) {
```

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -0,0 +1,623 @@
+#include "defs_urls.h"
+
+const int kDefaultRateRangePercent = 8;
+const double kRateDirectionInverted = -1;
```
Put this into an anonymous namespace.
Use constexpr

daschuer commented on src/preferences/dialog/dlgprefcontrols.cpp:
```diff
@@ -1,982 +0,0 @@
-        m_keyunlockModeControls.last()->set(m_keyunlockMode);
-    }
-
-    m_iNumConfiguredDecks = numdecks;
```
this line is lost.

Be-ing replied:
good catch

Be-ing committed:
use anonymous namespace

Be-ing committed:
restore lines lost when splitting up dlgprefcontrols.cpp

Be-ing committed:
fix tests that depend on rateRange

rateRange is a percent expressed as a decimal (0 - 1 range), but the
test previously set rateRange to 4.0. Commit
29ef9d1a708a1937dfaef90d2df33b43cdb718a6 changed rateRange to enforce a
limited range between 0.01 and 0.90, which broke the incorrect test
setup.

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into preferences_cleanup

Be-ing commented:
Notes addressed, tests fixed, and merge conflict fixed. I haven't done extensive manual testing yet.

Be-ing committed:
DlgPrefDecks: make Cancel button work

and pay down some technical debt

daschuer commented on src/engine/ratecontrol.h:
```diff
@@ -34,6 +34,31 @@ class RateControl : public EngineControl {
 
+    // Enumerations which hold the state of the pitchbend buttons.
+    // These enumerations can be used like a bitmask.
+    enum RATERAMP_DIRECTION {
```
normally we use camel case for enum names

Be-ing replied:
Yes, I just copy and pasted these without cleaning up the legacy code style yet.

Be-ing replied:
The amount of technical and design debt in the preferences is pretty overwhelming. The whole system could use a cleanup.

daschuer commented on src/engine/ratecontrol.h:
```diff
@@ -34,6 +34,31 @@ class RateControl : public EngineControl {
+    //  RATERAMP_RAMPBACK_NONE: returns back to normal all at once.
+    //  RATERAMP_RAMPBACK_SPEED: moves back in a linearly progresive manner.
+    //  RATERAMP_RAMPBACK_PERIOD: returns to normal within a period of time.
+    enum RATERAMP_RAMPBACK_MODE {
```
use camel case here as well

daschuer commented on src/engine/ratecontrol.h:
```diff
@@ -34,6 +34,31 @@ class RateControl : public EngineControl {
+    };
+
+    enum class RampMode {
+        Stepping = 0, // pitch takes a temporary step up/down a certain amount
```
Enum value names should be capital case

uklotzde replied:
Are you sure? I think we are using camel-case for enumerated values, at least for enum classes. Google allows both constant and macro style. IMHO the camel case constant style is more pleasant for reading.

Google also suggests to use a "k" prefix like for constants, but this is IMHO ridiculous for enum class values: RampMode::Stepping vs. RampMode::kStepping

uklotzde replied:
RampMode::STEPPING shows how the capitalized macro style breaks consistency, even more for literals with multiple terms.

uklotzde replied:
I've updated the coding guidelines, because an advice for enumerated values was missing. Feel free to discuss or change it I got this wrong.

https://mixxx.org/wiki/doku.php/coding_guidelines#forward_declared_strongly_typed_enums

There are still a lot of capitalized enum class enumerators in the code that need to be migrated at some time. I also used this style in the past.

uklotzde replied:
@daschuer Removed the "Bad" example. Using C-style enums indeed is still an option. The section should just show a typical example for the new C++-style enums that are the preferred way.

daschuer commented on src/engine/ratecontrol.h:
```diff
@@ -94,7 +125,7 @@ class RateControl : public EngineControl {
 
     // Values used when temp and perm rate buttons are pressed
-    static double m_dTemp, m_dTempSmall, m_dPerm, m_dPermSmall;
+    static double m_dTemporaryRateChangeCoarse, m_dTemporaryRateChangeFine, m_dPermanentRateChangeCoarse, m_dPermanentRateChangeFine;
```
This line is even for GitHub too long.
Please use an extra line and type for every declaration.

daschuer commented on src/engine/ratecontrol.h:
```diff
@@ -94,7 +125,7 @@ class RateControl : public EngineControl {
-    static double m_dTemp, m_dTempSmall, m_dPerm, m_dPermSmall;
+    static double m_dTemporaryRateChangeCoarse, m_dTemporaryRateChangeFine, m_dPermanentRateChangeCoarse, m_dPermanentRateChangeFine;
 
     ControlPushButton *buttonRateTempDown, *buttonRateTempDownSmall,
```
Could you take care here as well? 
```c++
ControlPushButton* buttonRateTempDown;
ControlPushButton* buttonRateTempDownSmall;
ControlPushButton* buttonRateTempUp;
ControlPushButton* buttonRateTempUpSmall;
```

daschuer commented:
I am sure, that we use capitals for plain enum values in the past.
I do not mind to change this for enum classes and also drop the k.
We only need to note this in our style guide.

daschuer commented:
> I've updated the coding guidelines, ...

Thank you @uklotzde . I do not like that bad example, because it is a good style for the old enums which pollutes the name space like macros. Using the old enums in new code is of cause bad. 
Do you agree? What is the best way around it? Just remove the bad example?

daschuer commented on src/test/golden_buffers/BasicProcessingTestPause:
```diff
@@ -1,512 +1,512 @@
-0.870045,0.870045
```
can you explain why the tests are failing after your changes?

Be-ing replied:
The explanation is in the commit message:

rateRange is a percent expressed as a decimal (0 - 1 range), but the
test previously set rateRange to 4.0. Commit
29ef9d1 changed rateRange to enforce a
limited range between 0.01 and 0.90, which broke the incorrect test
setup.

daschuer commented:
If I change to "Elapsed and Remaining" in the Skin quit and restart Mixxx, it is still at "Elapsed and Remaining" which we can considered as correct. But If I now go to the preferences and just click apply, the time changes to "Elapsed" even though the preferences option is at "Elapsed and Remaining"
Tested with Deere.

daschuer commented:
If I change the Cue mode, from Pioneer to Mixxx, the play button keeps blinking. It stops blinking after using it.

daschuer commented:
The numeric value of the ramping pitch band slider is somehow meaningless. 
It would be nice to set it to set the slider in relation to the fin an coarse values below. 
Can we mark the slider ends reasonable?

daschuer commented:
In 50% scaling is the Deere parallel waveforms are too high

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into preferences_cleanup

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into preferences_cleanup

Be-ing committed:
remove unnecesary file heading comments

Be-ing commented on src/preferences/dialog/dlgprefinterface.cpp:
```diff
@@ -0,0 +1,418 @@
+
+    QList<QDir> skinSearchPaths = m_pSkinLoader->getSkinSearchPaths();
+    QList<QFileInfo> skins;
+    for (QDir dir : skinSearchPaths) {
```
The line below requires that the variable is not const.

Be-ing committed:
break up long lines of code

Be-ing committed:
fix updating of track time display in preferences when loading dialog

Be-ing committed:
make DlgPrefInterface Cancel button work

Be-ing commented:
Ping. The rate range bug must be fixed for the beta release.

ronso0 commented:
Anything specific to test?

Be-ing commented:
Refer to the [checklist above](https://github.com/mixxxdj/mixxx/pull/1377#issuecomment-343722676)

ronso0 commented:
I started without a config file.
In all categories I checked, individual settings persist.

Sound Hardware: works

Library: works

Controllers: enabled mapping for USB mouse, setting persists. Restore Defaults won't uncheck it though

AutoDJ: works

Decks: Restore Defaults doesn't work, they appear to be applied in Preferences but aren't in skin. When going back to Preferences all my individual settings are back. Same after restart.

Interface: same issue

Waveforms: initially waveform zoom was set to 33% (started with Shade). Restore Defaults sets this to 20%, but individual setting Synchronize Zoom is always re-checked when opening Prefs again

Equalizer: deck-specific EQ/QuickEffect settings don't persist a restart, instead this is reset and individual settings from deck1 are applied

Crossfader: works

Live Broadcasting: don't now if it's intended, "Turn on Live..." is reset to Off after disconnecting Broadcast in the menu. No further testing

Normalization: Apply ReplayGain is re-checked after restart, replay gain mode persists

Recording: defaults don't persist after resetting from individual settings

Beat Detection: works

Key Detection: works

Be-ing commented:
Regarding restoring defaults, I was under the impression that clicking that button should not immediately apply the defaults and the Apply button should need to be pressed to get them to apply. @ronso0 are you saying that is not intuitive? Should Restore to Defaults also apply immediately?

ronso0 commented:
No, Restore Defaults sets all options and Apply ..well applies them unless one hits Cancel.
I saw the check list just after reporting my results, and considering the results, I didn't repeat the tests without Apply ;)

Be-ing commented:
So should we merge this as-is or should I make the Reset to Defaults buttons in the Interface and Decks panes immediately apply the defaults?

daschuer commented:
Reset do defaults should not imply Apply. A user may whish to just check the default value for a few items without applying all. 

How is the state of my findings? I think we need a Launchpad bugs if they are not addressed here.

Be-ing commented:
> If I change the Cue mode, from Pioneer to Mixxx, the play button keeps blinking. It stops blinking after using it.

I cannot reproduce this.

> The numeric value of the ramping pitch band slider is somehow meaningless.
It would be nice to set it to set the slider in relation to the fin an coarse values below.
Can we mark the slider ends reasonable?

With what? "Low" and "High"? Which label belongs on which side?

> In 50% scaling is the Deere parallel waveforms are too high

I'm not sure what you mean, but that doesn't sound like a bug with the preferences. It sounds like a bug in the waveforms or skin.

All other notes addressed.

daschuer commented:
Ah okay, so the PR related things are done. 
Good work, thank you!
